src/storage/src/render/upsert.rs

```diff
@@ -200,19 +212,80 @@ fn evaluate(
 }
 
+fn extract_kv<G: Scope>(
+    records: Collection<G, Row, Diff>,
+    key_indices_sorted: Vec<usize>,
+    key_indices: &[usize],
+) -> Collection<G, (Row, Row), Diff> {
+    let key_cols_are_sorted = &key_indices_sorted == key_indices;
+
+    let mut row_buf = Row::default();
+    // If the key columns are in order, we can pack them directly
+    // into `key_row_buf` while scanning the original row.
+    //
+    // Otherwise, we need to put them into `key_dv` and then invert the permutation
+    // before packing.
+    let mut key_row_buf = Row::default();
+    let mut key_dv = DatumVec::default();
+    let key_unsort_permutation = inverse_argsort(key_indices);
+    records.map(move |row| {
+        let mut row_packer = row_buf.packer();
+        let mut key_row_packer = key_row_buf.packer();
+        let values = &mut row.iter();
+        let mut next_idx = 0;
+        let mut key_dv = key_dv.borrow();
+        for &key_idx in key_indices_sorted.iter() {
+            // First, push the datums that are before `key_idx`
+            row_packer.extend(values.take(key_idx - next_idx));
+            // Then, add the key field to whichever buffer we're using for the key
+            let key_datum = values.next().unwrap();
+            if key_cols_are_sorted {
+                key_row_packer.push(key_datum);
+            } else {
+                key_dv.push(key_datum);
+            }
+            next_idx = key_idx + 1;
+        }
+        // Finally, push any columns after the last key index
+        row_packer.extend(values);
+
+        if !key_cols_are_sorted {
+            for &i in key_unsort_permutation.iter() {
+                key_row_packer.push(key_dv[i])
+            }
+        }
+        (key_row_buf.clone(), row_buf.clone())
+    })
+}
+
 /// Internal core upsert logic.
 fn upsert_core<G>(
     stream: &Stream<G, DecodeResult>,
     predicates: Vec<MirScalarExpr>,
     position_or: Vec<Option<usize>>,
     as_of_frontier: Antichain<Timestamp>,
     upsert_envelope: UpsertEnvelope,
-) -> Stream<G, (Result<Row, DataflowError>, u64, Diff)>
+    previous_ok: Collection<G, Row, Diff>,
+    mut previous_token: Option<Rc<dyn Any>>,
+) -> Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>
 where
     G: Scope<Timestamp = Timestamp>,
 {
-    let result_stream = stream.unary_frontier(
+    // Prepare sorted and structured `key_indices` required
+    // by the upsert operator, and a `DatumVec` used to avoid
+    // an allocation.
+    let mut key_indices_sorted = upsert_envelope.key_indices.clone();
+    key_indices_sorted.sort_unstable();
+
+    let previous_ok = extract_kv(
+        previous_ok,
+        key_indices_sorted.clone(),
+        &upsert_envelope.key_indices,
+    );
+    let result_stream = stream.binary_frontier(
+        &previous_ok.inner,
         Exchange::new(move |DecodeResult { key, .. }| key.hashed()),
+        Exchange::new(|((key, _v), _t, _r)| Some(Ok::<_, Infallible>(key)).hashed()),
```

Comment:
Should the `Infallible` here be a `DecodeError` to be sure that the hash functions match up? Afaict, `DecodeResult`'s `key` member is a `Option<Result<Row, DecodeError>>`.

Reply:
Good point! I was assuming that the hash of `Ok(x)` should be the same independent of the error type, but I suppose this isn't guaranteed.